Forced size command in math

sitandr opened the pull request:
Hello, I have recently seen #1117 and #1116, where to make a thing in fraction bigger there were suggested ugly constructions with boxes. It is quite common need if typing large math fractions and exponentials. I was a bit disgusted by currently available solutions to change the size type in math, so I thought it would be sensible to add to math library analogues of `\displaystyle` in LaTex. I made it work the same way as `bb`, `italic` and others. Though, I use there a string argument to select size instead of separate functions for each size.

Probably, it would be better to create a separate function for each size like it is done with `MathVariant`, but that seemed to me a bit of excessive for the start. If this feature *fits* the library, the proper names for these functions have to be selected as I'm still new to typst and don't know most of it's conventions.

> P.S. I'm sorry if I messed with something, that is actually my first PR.

PgBiel commented:
Hello, thanks for the PR! Please take a look at the following:

1. Please fix formatting for the CI to pass (run `cargo fmt` once in your codebase).
2. Please implement some tests, so we can see that the feature works as intended (and to catch possible future regressions). To do so, either edit some file in `tests/typ` or create one, in the appropriate category (in this case, `math`), following the format of other test files (note that each `---` is a separate test). Add enough tests to cover your additions, and then run them `cargo test --all --test tests`. Take a look at the generated png in `png/` - if the result is satisfactory (**and older tests weren't affected**), then you may accept the generated png as the new reference png (expected result) by prefixing the test command with `UPDATE_EXPECT=1`.

Thanks!

PgBiel commented:
Regarding the PR's contents, my opinions thus far:

1. I think the name `withsize` could be improved; perhaps `sized`? Not sure here.
2. I think a better syntax (ignoring the names here) would be `withsize.display(...)`, `withsize.inline(...)` etc., instead of `withsize("display", ...)`, `withsize("inline", ...)` etc. I was thinking that you could use function scopes for that, but, on second thought, `withsize` could just be a regular module with those functions inside. (Doesn't mean you have to duplicate your code - you can keep a separate _Rust_ function which you call for each of those sizes.) Separating functions like that would make it easier to e.g. diagnose errors and improve autocomplete.
3. I wonder if all those sizes are necessary, or if just adding support for display is enough?

sitandr commented:
Thanks for all yours advices! I use now `sized` as module as you have suggested. The names are currently the same names as they are in MathSize enum, with exception of "text" size replaced by "inline" to avoid confusion (not sure it is a good idea, though). So now it is like `$sized.display(A/B), sized.inline(A/B), sized.script(A/B), sized.scriptscript(A/B)$` (I added these and some others to test).

Answering concern about needs of other set-sizes than display, as a LaTeX user, I sometimes really need to use `\scriptstyle` and `\textstyle` commands to type complicated powers to collapse non-vital things into something smaller. They are used a bit less than `\displaystyle`, but they are helpful to get the look you want.  I'm sure that there will be people who lack these size manipulators. All in all, it would be strange to have only `display` and no others.

I decided to use duplicating code to write these commands as it is done with MathVariant. There are less different sizes than variants in MathVariant, so I decided to keep the code more uniform (also, most of code actually consists of the doc strings that will be written in any case).

PgBiel commented on library/src/math/style.rs:
````diff
@@ -178,6 +178,84 @@ pub fn bb(
+///
+/// ## Example
+/// ```example
+/// $ script.text(sum_i (x_i mu_i)/sigma_i^2)/(sum_i (mu_i^2)/sigma_i^2) $
````
```suggestion
/// $ sized.script.text(sum_i (x_i mu_i)/sigma_i^2)/(sum_i (mu_i^2)/sigma_i^2) $
```

PgBiel commented on library/src/math/style.rs:
````diff
@@ -178,6 +178,84 @@ pub fn bb(
+///
+/// ## Example
+/// ```example
+/// $ scriptscript.text(sum_i (x_i mu_i)/sigma_i^2)/(sum_i (mu_i^2)/sigma_i^2) $
````
```suggestion
/// $ sized.scriptscript.text(sum_i (x_i mu_i)/sigma_i^2)/(sum_i (mu_i^2)/sigma_i^2) $
```

PgBiel commented on library/src/math/style.rs:
```diff
@@ -178,6 +178,84 @@ pub fn bb(
+        .into()
+}
+
+/// Forced script-script style in math.
```
The name "script-script" isn't (IMHO) that intuitive (this isn't your fault!) - perhaps you could add a brief explanation to the docs that this corresponds to (e.g.) a smaller script size than "script"?

PgBiel commented:
@sitandr thanks for explaining everything! The PR is looking much better now (at least in terms of shape and guidelines met). I've added a few quick comments above, but they're nothing much. Thanks once again!

laurmaedje commented:
To be honest, I would not have a module here. This is very similar to `bold`/`italic` or `bb`/`cal` etc. and it should be handled consistently. The names _are_ a bit less self-explanatory and usage rarer than for the other two though, so I'm not 100% sure on this.

Should we do it, we also need an entry in `docs/src/reference/groups.yml` so that the documentation correctly groups them together into a single page.

laurmaedje commented:
We could also stray away from TeX and use clearer names.

sitandr commented:
I thought a lot about alternative names and didn't really succed. Tex is standart in math typesetting, so we need to make something familiar, using names just out of nowhere is strange. I decided we could replace `scriptscript` with a shorter `sscript` (second-script or script-script). So the names will be changed as

| Tex | Typst |
|----|--------|
| `\displaystyle` | `display` |
| `\textstyle` | `inline` |
| `\scriptstyle` | `script` |
| `\scriptscriptstyle` | `sscript` |

I removed module for these names as they are actually quite good on their own. It will be intuitive for math users that `display` and `script` refers to style, not some embed content. `inline` could otherwise refer to creating equation from code, but there is already `equation` with `block` arg, so it is quite intuitive too. I don't think these names in math could be confused with something else.

Additionally, I added named `cramp` arg to all these functions enabled by default for scripts to make it more script-like (users ususall). It doesn't change much (the difference is about 10%), but it is an expected thing to have as we have "cramped" setting for mathstyle.

laurmaedje commented:
Thanks!
